enhancement: Add events for starting, stopping, and reloading

ktff opened the pull request:
Closes #3448.

Added events emit:
- `Vector is starting.`
- `Vector is reloading.`
- `Vector is stopping.`

logs respectively. This differs from what it used to be, but it reflects the metrics and is more streamlined.

### Open questions

- [x] logs are emitted under fixed `target = "vector"` to avoid `vector::internal_events::process:` being logged as the call site. Haven't seen what's our stance on that. It should be possible to extend `emit!` macro to set target as the actual emit site, or is this the intended behavior? (Edit: it should be fine to use target for now)

- [x] Should we use lowercase `vector` ? (Edit: use `Vector` for emphasis) 

<!--
**Your PR title must conform to the conventional commit spec!**

  <type>!?(<scope>): <description>

  * `type` = chore, docs, enhancement, newfeat, perf
  * `!` = signals a breaking change
  * `scope` = https://github.com/timberio/vector/blob/master/.github/semantic.yml#L4
  * `description` = short description of the change

Examples:

  * enhancement(file source): Added `sort` option to sort discovered files
  * feat(new source): Initial `statsd` source
  * fix(file source): Fixed a bug discovering new files
  * perf(observability): Improved logging performance
  * docs: Clarified `batch_size` option
-->

ktff committed:
Add process events

Signed-off-by: ktf <krunotf@gmail.com>

jamtur01 commented on src/main.rs:
```diff
@@ -169,9 +165,10 @@ fn main() {
             }
         };
 
+        emit!(VectorStopped);
```
Obviously very much new here ... for consistency should the `emit!` for shutdown be where the `info!` is currently?

ktff replied:
Yea, I thought this was better place, but this clearer.

lukesteensen reviewed (changes requested):
Looking at this again and taking @jamtur01's comment into account, I'm starting to think that it's worth differentiating between normal graceful shutdowns and the immediate forced shutdown. That would clear up the code path confusion and give us higher quality information about what's happening with the process.

binarylogic commented on src/internal_events/process.rs:
```diff
@@ -0,0 +1,63 @@
+    fn emit_logs(&self) {
+        info!(
+            target: "vector",
+            message = "Vector is reloading.",
```
```suggestion
            message = "Vector has reloaded.",
```

binarylogic commented on src/internal_events/process.rs:
```diff
@@ -0,0 +1,63 @@
+    fn emit_logs(&self) {
+        info!(
+            target: "vector",
+            message = "Vector is starting.",
```
```suggestion
            message = "Vector started.",
```

ktff replied:
Maybe `Vector has started.` so it has the scheme as other messages?

binarylogic replied:
Yep, that works as well.

binarylogic commented on src/internal_events/process.rs:
```diff
@@ -0,0 +1,63 @@
+    fn emit_logs(&self) {
+        info!(
+            target: "vector",
+            message = "Vector is stopping."
```
```suggestion
            message = "Vector has stopped."
```

I know this one is strange since Vector wouldn't be able to log if it was technically stopped, but it should still happen after all components have stopped. I also want to keep our past-tense theme.

ktff committed:
Use past tense

Signed-off-by: Kruno Tomola Fabro <krunotf@gmail.com>

ktff committed:
One last metric capture

Signed-off-by: Kruno Tomola Fabro <krunotf@gmail.com>

ktff committed:
One more move

Signed-off-by: Kruno Tomola Fabro <krunotf@gmail.com>

ktff commented:
Going with the theme of past tense, the points of emissions have been moved to reflect this.

---

Regarding the shutdown event, I've added a fix to the `internal_metrics` source because ,as it is, we won't capture it,almost for certain, before we shutdown the source. So I've added a fix for it. 

@lukesteensen this ties into immediate forced shutdown, it is highly likely that we won't propagate any emitted event on that path further than `internal_metrics` source even with the fix.  

---

There are also errors in main which we could turn into events.

ktff committed:
Remove unused use

Signed-off-by: Kruno Tomola Fabro <krunotf@gmail.com>
